feat: Scope KMS policy to the exact KMS key

arya-girish-k opened the pull request:
### Description
Scope KMS policy to the exact KMS key
[Git_issue.](https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/issues/223)
<!--- Replace this text with a summary of the changes in this PR. Include why the changes are needed and context about the changes. List required dependencies. If there is a Git issue for the change, please link to it. --->

### Release required?
<!--- Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning). --->

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content
As part of addressing this issue,made the following modification:

1. Updated the KMS policy and added an additional policy for the HPCS key in the root module and DA.
2. Introduced a boolean variable `is_hpcs_key` to create the second policy.
3. Implemented a CRN parser to retrieve the `kms_instance_guid` from `kms_key_crn_parser`, removing the `existing_kms_instance_guid`.
4. Updated the validation

<!--- If a release is required, replace this text with information that users need to know about the release. Write the release notes to help users understand the changes, and include information about how to update from the previous version.

Your notes help the merger write the commit message for the PR that is published in the release notes for the module. --->

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

ocofaigh commented:
@arya-girish-k any update on this?

ocofaigh commented on main.tf:
```diff
@@ -52,34 +52,97 @@ resource "ibm_resource_instance" "secrets_manager_instance" {
-  description                 = "Allow all Secrets Manager instances in the resource group ${var.resource_group_id} to read from the ${local.kms_service_name} instance GUID ${var.existing_kms_instance_guid}"
+  target_resource_instance_id = local.kms_instance_guid
+  roles                       = ["Viewer"]
+  description                 = "Allow all Secrets Manager instances in the resource group ${var.resource_group_id} to read from the ${local.kms_service_name} instance GUID ${local.kms_instance_guid}."
```
`Allow all Secrets Manager instances in the resource group ${var.resource_group_id} viewer access to the ${local.kms_service_name} instance GUID ${local.kms_instance_guid}.`

arya-girish-k replied:
[fixed](https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/pull/295/files#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR154)

ocofaigh reviewed (changes requested):
Lets forget about DA updates in this PR - the DA is being refactored in https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/pull/300 and it will handle the DA updates

ocofaigh reviewed (changes requested):
see comments

arya-girish-k commented on main.tf:
```diff
@@ -52,34 +52,97 @@ resource "ibm_resource_instance" "secrets_manager_instance" {
-resource "ibm_iam_authorization_policy" "kms_policy" {
-  count                       = var.kms_encryption_enabled && !var.skip_kms_iam_authorization_policy && var.existing_sm_instance_crn == null ? 1 : 0
+# Create auth policy (scoped to exact KMS key)
+resource "ibm_iam_authorization_policy" "secrets_manager_kms_policy" {
```
[Fixed](https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/pull/295/files#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR99)

arya-girish-k commented on main.tf:
```diff
@@ -43,7 +41,7 @@ resource "ibm_resource_instance" "secrets_manager_instance" {
   parameters = {
     "allowed_network" = var.allowed_network
-    "kms_instance"    = var.existing_kms_instance_guid
+    "kms_instance"    = var.kms_instance_guid
```
[fixed](https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/pull/295/files#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR46)

arya-girish-k commented on main.tf:
```diff
@@ -71,33 +69,96 @@ resource "ibm_iam_authorization_policy" "iam_groups_policy" {
 
 # workaround for https://github.com/IBM-Cloud/terraform-provider-ibm/issues/4478
-resource "time_sleep" "wait_for_authorization_policy" {
+resource "time_sleep" "wait_for_sm_kms_authorization_policy" {
```
fixed it.

arya-girish-k commented on outputs.tf:
```diff
@@ -37,5 +37,8 @@ output "secrets" {
-
+output "kms_key_crn" {
+  value       = var.kms_key_crn
+  description = "List of secret mananger secret config data"
```
It was added as part of testing, but has now been reverted.

arya-girish-k commented on solutions/standard/outputs.tf:
```diff
@@ -32,3 +32,7 @@ output "secrets_manager_region" {
 }
+output "kms_key_crn" {
+  value       = local.kms_key_crn
+  description = "List of secret mananger secret config data"
```
It was added as part of testing, but has now been reverted.

arya-girish-k commented on solutions/standard/main.tf:
```diff
@@ -28,36 +28,108 @@ locals {
+}
+
+# Create auth policy (scoped to exact KMS key)
+resource "ibm_iam_authorization_policy" "secrets_manager_kms_policy" {
```
[fixed](https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/pull/295/files#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR99)

arya-girish-k commented on main.tf:
```diff
@@ -71,33 +69,96 @@ resource "ibm_iam_authorization_policy" "iam_groups_policy" {
-    )
-  ) : null
+  create_kms_auth_policy  = var.kms_encryption_enabled && !var.skip_kms_iam_authorization_policy && var.existing_sm_instance_crn == null
+  create_hpcs_auth_policy = local.create_kms_auth_policy == true && local.kms_service_name == "hs-crypto" ? 1 : 0
```
As per the [discussion](https://ibm-cloudplatform.slack.com/archives/CG06FQ31P/p1742310055504799) added new boolean variable `is_hpcs_key` .

arya-girish-k commented on main.tf:
```diff
@@ -6,11 +6,9 @@
   validate_kms_vars = var.kms_encryption_enabled && var.kms_key_crn == null && var.existing_sm_instance_crn == null ? tobool("When setting var.kms_encryption_enabled to true, a value must be passed for var.kms_key_crn") : true
   # tflint-ignore: terraform_unused_declarations
-  validate_auth_policy = var.kms_encryption_enabled && var.skip_kms_iam_authorization_policy == false && var.existing_kms_instance_guid == null && var.existing_sm_instance_crn == null ? tobool("When var.skip_kms_iam_authorization_policy is set to false, and var.kms_encryption_enabled to true, a value must be passed for var.existing_kms_instance_guid in order to create the auth policy.") : true
+  validate_auth_policy = var.kms_encryption_enabled && var.skip_kms_iam_authorization_policy == false && var.kms_key_crn == null && var.existing_sm_instance_crn == null ? tobool("When var.skip_kms_iam_authorization_policy is set to false, and var.kms_encryption_enabled to true, a value must be passed for var.kms_key_crn in order to create the auth policy.") : true
```
> The value of `skip_kms_iam_authorization_policy` should not even be used in any validation logic here. If `kms_encryption_enabled` is set to true, then a value is needed for either `kms_key_crn`, regardless of what value `skip_kms_iam_authorization_policy` is

```validate_kms_values = (!var.kms_encryption_enabled && var.kms_key_crn != null && var.existing_sm_instance_crn == null) ? tobool("When passing values for var.kms_key_crn, you must set 'kms_encryption_enabled' to true. Otherwise set 'kms_encryption_enabled' to false to use default encryption") : true```
Updated the existing validation.

arya-girish-k commented:
1.Added two policies in the rootmodule and DA.
2.Added crn parser ,since retriving kms_instance_guid from kms_key_crn_parser,removed existing_kms_instance_guid.
3.Added boolean variable `is_hpcs_key` to create the second policy
4.Faced the below  error while testing:
```summary: |
│   CreateNotificationsRegistrationWithContext failed Invalid authorization with Event Notifications instance 
```
Since it is [Known](https://ibm-cloudplatform.slack.com/archives/CG06FQ31P/p1741711327690389) error, removed this [line](https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/pull/295/files#diff-07f337e38ed1996f30ee4f04357ab23318c254cb606aaf4745d564cbd546722dL35) in complete example.

arya-girish-k commented:
/run pipeline

arya-girish-k commented:
/run pipeline

ocofaigh commented on variables.tf:
```diff
@@ -61,18 +61,16 @@ variable "skip_kms_iam_authorization_policy" {
 
+variable "is_hpcs_key" {
+  type        = bool
+  description = "Set it to true if the key is Hyper Protect Crypto Services key"
```
I would default this to false as most consumers probably use Key Protect. The variable description should reference the `kms_key_crn` inpout

ocofaigh commented on solutions/standard/variables.tf:
```diff
@@ -192,6 +192,12 @@ variable "existing_kms_instance_crn" {
 
+variable "is_hpcs_key" {
+  type        = bool
+  default     = true
```
default this to false

ocofaigh commented on modules/fscloud/variables.tf:
```diff
@@ -50,17 +50,15 @@ variable "skip_kms_iam_authorization_policy" {
 
+variable "is_hpcs_key" {
+  type        = bool
+  description = "Set to true if the key is hpcs, otherwise false."
```
default this to false

ocofaigh commented on examples/fscloud/main.tf:
```diff
@@ -52,14 +52,14 @@ module "event_notification" {
+  region                   = var.region
+  secrets_manager_name     = "${var.prefix}-secrets-manager" #tfsec:ignore:general-secrets-no-plaintext-exposure
+  sm_tags                  = var.resource_tags
+  is_hpcs_key              = true
```
You should use a CRN parser to determine if the value of `kms_key_crn` is HPCS as its possible a user could use a key protect key when calling this example. It should not cause any issue with "value not known until apply" since its always going to be a required input to the example and the key will never be created in this example

ocofaigh commented on examples/basic/main.tf:
```diff
@@ -13,4 +13,5 @@ module "secrets_manager" {
   secrets_manager_name = "${var.prefix}-secrets-manager" #tfsec:ignore:general-secrets-no-plaintext-exposure
   sm_service_plan      = "trial"
   sm_tags              = var.resource_tags
+  is_hpcs_key          = true
```
remove this from the basic example

ocofaigh reviewed (changes requested):
see final comments

Aayush-Abhyarthi commented:
/run pipeline

ocofaigh commented on solutions/standard/variables.tf:
```diff
@@ -192,6 +192,12 @@ variable "existing_kms_instance_crn" {
+variable "is_hpcs_key" {
+  type        = bool
+  default     = false
+  description = "Set it to true if the key provided through the `existing_kms_instance_crn` is Hyper Protect Crypto Services key"
```
You know what, we actually don't need this input in the DA. We can use the CRN parser to check if its KP or HPCS by parsing either `existing_kms_instance_crn` (when creating new key) or `existing_secrets_manager_kms_key_crn` when using existing key.

We only need it in the module because its possible user could be creating the KP instance and/or key in same terraform state. But thats not the case in the DA. The `existing_kms_instance_crn` or `existing_secrets_manager_kms_key_crn` will always be known before apply time

ocofaigh reviewed (changes requested):
just realised we don't need the new boolean in the DA - we can parse the CRN (see comment)

Aayush-Abhyarthi commented:
/run pipeline

Aayush-Abhyarthi commented:
/run pipeline

ocofaigh reviewed (changes requested):
We seem to be missing logic in the module that verifies if `is_hpcs_key` is set to true, it matches the instance type in the `kms_key_crn` value. Can you add it please?

Aayush-Abhyarthi commented:
/run pipeline

terraform-ibm-modules-ops commented:
:tada: This PR is included in version 1.26.0 :tada:

The release is available on:
- [GitHub release](https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/releases/tag/v1.26.0)
- `v1.26.0`

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
